### PR TITLE
Numerous optimisations

### LINKED ISSUE #3417: More Text = Less Frames
So for the record I'm running Win7 64 bit, 8GB of RAM, an ATI HD7870 and an AMD Phenom II X4 Black Edition @ 3.4Ghz.

Whenever text is rendered on screen I lose frames like crazy, to the point where opening autopilot can dip my framerate from a perfect 60 to sub-15. I'm not sure what's causing this but I have the latest proprietary drivers from ATI installed as well. Other than this the game runs flawlessly.

On latest version: 20150701

### PR DESCRIPTION
**Pull request by fluffyfreak**
A collection of optimisations done to fix #3417 aka "More Text = Less Frames".

- NavLights were taking a ridiculously long time to update due to stupidity.
- Don't update the time text unless the time has changed.
- Reserve space for text before filling out the buffers.
- If the text and colour hasn't changed then don't update the buffers.
- The big one: Cache the vertex buffers created by the TextureFont and reuse them.
- Avoid rebuilding the line rendering multiple times per-frame in SectorView!

On my laptop this improves the SectorView (F2) screen from approximately 17fps with all text on show to a steady 60fps.

Andy

### CODE REVIEW

**Comment by impaktor** [comment]
Sweet. :+1:

**Comment by fluffyfreak** [comment]
Yeeeees, tricky.

**Comment by fluffyfreak** [comment]
@impaktor sorted! :)

**Comment by impaktor** [comment]
Hm, why doesn't using `Pi::game->GetTime() : 0.0;` work? Isn't that the preferred way of getting the time? Is it some Makefile related change that needed to be made due to the added includes?

Either way, feels like I'm talking out my bum. Just thinking out loud.

**Comment by walterar** [comment]
pioneer: NavLights.cpp:189: virtual void NavLights::LoadFromJson(const Json::Value&): Assertion `lightIndex < glit->m_lights.size()' failed.

**Comment by fluffyfreak** [comment]
Ah yes, this is probably going to require a save game bump :(

Or I will have to handle the old save format safely. Suggestions?

**Comment by walterar** [comment]
The old saved games not working. The trouble is that new saved games either.

**Comment by fluffyfreak** [comment]
@walterar ok that works fine for me now, doesn't assert even in full debug.
Also found some more optimisations for rendering the SectorView.

**Comment by walterar** [comment]
Start in London
Save Game "London"
leave
Load Game "London"

ver git (8aea83a) on: Linux

System Name: Linux
Host Name: oper-H61M-USB3-B3
Release(Kernel) Version: 3.2.0-87-generic
Kernel Build Timestamp: #125-Ubuntu SMP Fri Jun 19 08:25:10 UTC 2015
Machine Arch: x86_64
Domain Name: (none)

SDL Version 2.0.3
Initialized OpenGL 3.1, with extensions, renderer
invalid axis binding '' in config file for BindAxisPitch
invalid axis binding '' in config file for BindAxisRoll
invalid axis binding '' in config file for BindAxisYaw
started 4 worker threads
ShipType::Init()
Lua::Init()
Server agent disabled
GalaxyGenerator::Init()
Creating new galaxy generator 'legacy' version 1
Number of factions added: 103
StarSystemCache: misses: 2227, slave hits: 0, master hits: 0
SectorCache: misses: 1571, slave hits: 0, master hits: 11138
FaceParts::Init()
Face Generation source images loaded.
new ModelCache
Shields::Init
BaseSphere::Init
CityOnPlanet::Init
CreateCollisionMesh for : (kbuilding01)
CreateCollisionMesh for : (kbuilding02)
CreateCollisionMesh for : (kbuilding03)
CreateCollisionMesh for : (newbuilding1)
CreateCollisionMesh for : (newbuilding10)
CreateCollisionMesh for : (newbuilding11)
CreateCollisionMesh for : (newbuilding2)
CreateCollisionMesh for : (newbuilding3)
CreateCollisionMesh for : (newbuilding4)
CreateCollisionMesh for : (newbuilding5)
CreateCollisionMesh for : (newbuilding6)
CreateCollisionMesh for : (newbuilding7)
CreateCollisionMesh for : (newbuilding8)
CreateCollisionMesh for : (newbuilding9)
Got 14 buildings of tag city_building
 - kbuilding01: 29.664692
 - kbuilding02: 46.268344
 - kbuilding03: 41.542333
 - newbuilding1: 36.202756
 - newbuilding10: 16.297418
 - newbuilding11: 34.961034
 - newbuilding2: 22.822782
 - newbuilding3: 22.649894
 - newbuilding4: 25.978193
 - newbuilding5: 22.908506
 - newbuilding6: 29.817877
 - newbuilding7: 22.674426
 - newbuilding8: 23.154979
 - newbuilding9: 36.528138
End of buildings.
SpaceStation::Init
CreateCollisionMesh for : (ground_station)
ground_station has:
 14 entrances,
 14 pads,
 0 exits
CreateCollisionMesh for : (new_ground)
new_ground has:
 6 entrances,
 6 pads,
 0 exits
CreateCollisionMesh for : (orbital_station_2-10k)
orbital_station_2-10k has:
 5 entrances,
 64 pads,
 5 exits
CreateCollisionMesh for : (orbital_station_2-2k)
orbital_station_2-2k has:
 4 entrances,
 18 pads,
 4 exits
CreateCollisionMesh for : (orbital_station_2-5k)
orbital_station_2-5k has:
 5 entrances,
 64 pads,
 5 exits
CreateCollisionMesh for : (orbital_station_2-5k10k)
orbital_station_2-5k10k has:
 5 entrances,
 64 pads,
 5 exits
NavLights::Init
Sfx::Init
Sound::Init


Loading took: 29412.852174 milliseconds
CreateCollisionMesh for : (ac33)
CreateCollisionMesh for : (amphiesma)
CreateCollisionMesh for : (deneb)
CreateCollisionMesh for : (dsminer)
CreateCollisionMesh for : (kanara_civ)
CreateCollisionMesh for : (lunarshuttle)
CreateCollisionMesh for : (malabar)
CreateCollisionMesh for : (molamola)
CreateCollisionMesh for : (natrix)
CreateCollisionMesh for : (nerodia)
CreateCollisionMesh for : (pumpkinseed)
CreateCollisionMesh for : (sinonatrix)
CreateCollisionMesh for : (storeria)
CreateCollisionMesh for : (varada)
CreateCollisionMesh for : (vatakara)
CreateCollisionMesh for : (venturestar)
CreateCollisionMesh for : (wave)
CreateCollisionMesh for : (xylophis)
Game::LoadGame('London')
savefile version: 81
Creating new galaxy generator 'legacy' version 1
Clearing and re-using previous Galaxy object
StarSystemCache: misses: 0, slave hits: 0, master hits: 0
SectorCache: misses: 101, slave hits: 0, master hits: 1
Sol:
    height fractal: Ellipsoid
    colour fractal: StarG
    seed: 1685209112
Mercury:
    height fractal: HillsCraters2
    colour fractal: Rock
    seed: 6
Venus:
    height fractal: RuggedLava
    colour fractal: Desert
    seed: 1710384323
Earth:
    height fractal: Mapped
    colour fractal: EarthLikeHeightmapped
    seed: 1717547535
NavLights group (0) added
NavLights group (1) added
NavLights group (2) added
NavLights group (3) added
NavLights group (4) added
NavLights group (5) added
NavLights group (6) added
pioneer: NavLights.cpp:189: virtual void NavLights::LoadFromJson(const Json::Value&): Assertion `lightIndex < glit->m_lights.size()' failed.

OpenGL version 3.1.0 NVIDIA 331.113, running on NVIDIA Corporation GeForce GTX 650/PCIe/SSE2
Available extensions:
Shading language version: 1.40 NVIDIA via Cg compiler

**Comment by fluffyfreak** [comment]
Solved by completely removing the NavLight state saving/loading for individual lights which are reset and updated per-frame anyway making saving or loading them pointless.

**Comment by walterar** [comment]
Now it works! :+1: 

I wonder, how many things are saved in vain? :(

I will continue by testing this. If I have something new, I will tell you here. For now, you can sleep peacefully. :))

**Comment by fluffyfreak** [comment]
@impaktor the game->GetTime didn't work due to where it was being used not depending on Pi or Game at link time I think. Worked out for the best anyway since I wanted a system time not the games idea of time.

I'm done with this now I just had a final change to make to do with filling up the line drawing buffers in SectorView.

That screen really needs converting to the new UI.

**Comment by impaktor** [comment]
OK. 

The fix savegame loading issue is nothing that needs/should be removed on the next save bump?

If this is done, I guess we give @walterar a chance to test it again, before merge.

**Comment by fluffyfreak** [comment]
Actually I think it won't need a save game bump, I removed the light state loading/saving code so it should load the old save games just fine. It will just ignore the light states and then fix them on first update.

**Comment by SpencerBelleau** [comment]
Thanks man, I really appreciate it. Now I've just gotta get this compiled.

**Comment by fluffyfreak** [comment]
@SpencerBelleau no problems, it should be available in the next build so you can just wait until that's done :)

File: src/text/TextureFont.cpp [hunk 1]
```diff
@@ -282,28 +296,91 @@ Color TextureFont::PopulateMarkup(Graphics::VertexArray &va, const std::string &
...
+	}
+
+	// update the last access time
+	const double lastAccessTime = 0.001 * double(SDL_GetTicks());
+	found->second.first = lastAccessTime;
+
+	if ((lastAccessTime - m_lfLastCacheCleanTime) > CACHE_EVICTION_TIME) {
+		const Uint32 uDeleted = CleanVertexBufferCache();
+		Output("NumDeleted %d\n", uDeleted);
+		m_lfLastCacheCleanTime = lastAccessTime;
```
Comment (impaktor) [inline src/text/TextureFont.cpp]: @fluffyfreak this sure is an active debug output, flooding my terminal.
```
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
NavLights group (0) added
ALSA lib pcm.c:7810:(snd_pcm_recover) underrun occurred
NumDeleted 0
NumDeleted 0
ALSA lib pcm.c:7810:(snd_pcm_recover) underrun occurred
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 3
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
NumDeleted 0
```

**Comment by impaktor** [comment]
I don't know if this warning is to be taken seriously, but reporting it here:

```
GuiTextLayout.cpp: In member function ‘void Gui::TextLayout::Update(float, const Color&)’:
GuiTextLayout.cpp:107:17: warning: comparing floating point with == or != is unsafe [-Wfloat-equal]
```

**Comment by fluffyfreak** [comment]
Sorted :)

**Comment by impaktor** [comment]
Thanks!

Today must have been the most active pioneer-day in 6 months.

**Comment by impaktor** [comment]
@SpencerBelleau A new build has just been started, and should be done in the next few hours. Make sure to check the pioneer web site. So no compilation is required on your part.

**Comment by SpencerBelleau** [comment]
So uhh, turns out it did nothing for me at all. Same issues, same framerate. Sorry about that.

**Comment by fluffyfreak** [comment]
NO problem, I've re-opened issue #3417 to track any future progress. It was worthwhile anyway as I did fix a whole bunch of really stupid code :)

**Pull request merged**